File: dotnet/src/Frontend/src/widgets/Log.js
import React, { Component } from 'react'
import {Table, Column, Cell} from 'fixed-data-table'
import "fixed-data-table/dist/fixed-data-table.css"
import { map } from "../Util.ts"

const SortTypes = {
  ASC: 1,
  DESC: -1
};

class Sort {
  /**
   * @param {string} column
   * @param {number} direction
   */
  constructor(column, direction) {
    this.column = column;
    this.direction = direction;
  }

  equals(sort) {
    return sort instanceof Sort
      ? this.column === sort.column && this.direction === sort.direction
      : false;
  }
}

const SortableCell = props => {
  const col = typeof props.children === "string" ? props.children : "";
  return (
    <Cell style={{cursor: "pointer"}}
      onClick={ev => props.onClick(col)}
    >{col + (props.sort && col === props.sort.column ?
      (props.sort.direction === SortTypes.ASC ? '↓' : '↑') : "")}</Cell>
  );
}

class View extends Component {
  constructor(props) {
    super(props);
    var logs = props.global.state.logs;
    this.state = {
      filter: null,
      logLevel: "none",
      sort: null,
      columns: {
        LogLevel: true,
        Time: true,
        Tag: true,
        Tier: true
      },
      logs: logs,
      viewLogs: logs
    };
    IrisLib.subscribe(props.model.observable, kv => {
      if (kv.key === "filter") {
        this.updateViewLogs({filter: kv.value})
      }
      if (kv.key === "logLevel") {
        this.updateViewLogs({logLevel: kv.value})
      }
      else if (kv.key.startsWith("columns/")) {
        let key = kv.key.replace("columns/","")
        this.setState({columns: Object.assign(this.state.columns, {[key]: kv.value})});
      }
    });
  }

  updateViewLogs({
    logs = this.state.logs,
    filter = this.state.filter,
    logLevel = this.state.logLevel,
    sort = this.state.sort
  }) {
    let viewLogs = logs;
    if (filter) {
      let reg = new RegExp(filter, "i");
      viewLogs = viewLogs.filter(log => reg.test(log.Message));
    }
    if (logLevel && logLevel !== "none") {
      viewLogs = viewLogs.filter(log => IrisLib.toString(log.LogLevel) === logLevel);
    }
    if (sort) {
      viewLogs = viewLogs.sort((log1,log2) => {
        let col1 = IrisLib.toString(log1[sort.column]), col2 = IrisLib.toString(log2[sort.column]);
        var res = col1 < col2 ? 1 : (col1 === col2 ? 0 : -1);
        return sort.direction === SortTypes.ASC ? res : res * -1;
      })
    }
    this.setState({filter, sort, logLevel, logs, viewLogs})
  }

  componentDidMount() {
    this.disposable =
      this.props.global.subscribe("logs", logs =>
        this.updateViewLogs({logs}));
  }

  componentWillUnmount() {
    if (this.disposable) {
      this.disposable.dispose();
    }
  }

  onSorted(col) {
    let sort = this.state.sort;
    if (sort instanceof Sort && sort.column === col) {
      sort = new Sort(col, sort.direction * -1);
    }
    else {
      sort = new Sort(col, SortTypes.ASC);
    }
    this.updateViewLogs({sort});
  }

  render() {
   const onSorted = this.onSorted.bind(this);
   return (
    <Table
      rowsCount={this.state.viewLogs.length}
      rowHeight={30}
      headerHeight={30}
      width={700}
      height={600}
      >
        {this.state.columns.LogLevel ?
          <Column width={70}
            header={<SortableCell sort={this.state.sort} onClick={onSorted}>LogLevel</SortableCell>}
            cell={({rowIndex, ...props}) => (
              <Cell {...props}>{IrisLib.toString(this.state.viewLogs[rowIndex].LogLevel)}</Cell>
          )} /> : null}
        {this.state.columns.Time ?
          <Column width={80}
            header={<SortableCell sort={this.state.sort} onClick={onSorted}>Time</SortableCell>}
            cell={({rowIndex, ...props}) => (
              <Cell {...props}>{new Date(this.state.viewLogs[rowIndex].Time).toLocaleTimeString()}</Cell>
          )} /> : null }
        {this.state.columns.Tag ?
          <Column width={75}
            header={<SortableCell sort={this.state.sort} onClick={onSorted}>Tag</SortableCell>}
            cell={({rowIndex, ...props}) => (
              <Cell {...props}>{this.state.viewLogs[rowIndex].Tag}</Cell>
          )} /> : null }
        {this.state.columns.Tier ?
          <Column width={75}
            header={<SortableCell sort={this.state.sort} onClick={onSorted}>Tier</SortableCell>}
            cell={({rowIndex, ...props}) => (
              <Cell {...props}>{IrisLib.toString(this.state.viewLogs[rowIndex].Tier)}</Cell>
          )} /> : null }
        <Column width={400}
          header={<Cell>Message</Cell>}
          cell={({rowIndex, ...props}) => (
            <Cell {...props} style={{whiteSpace: "nowrap"}}>
              {this.state.viewLogs[rowIndex].Message}
            </Cell>
        )} />
      </Table>
    )
  }
}

class TitleForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      filter: "",
      logLevel: "none",
      setLogLevel: "debug",
      columns: {
        LogLevel: true,
        Time: true,
        Tag: true,
        Tier: true
      }
    };
  }

  renderDropdown(title, values, generator) {
    return (
    <div className="iris-dropdown">
    <div className="iris-dropdown-button">{title}</div>
    <div className="iris-dropdown-content">
      {values.map(x => <div key={x}>{generator(x)}</div>)}
    </div>
  </div>
  )}

  render() {
    return (
    <div>
      <input type="text" placeholder="Filter by regex..."
        value={this.state.filter} onChange={ev => {
          this.props.observable.trigger({key: "filter", value: ev.target.value});
          this.setState({filter: ev.target.value});
        }} />
      {this.renderDropdown("Columns", ["LogLevel", "Time", "Tag", "Tier"], col => (
        <label>
          <input type="checkbox" checked={this.state.columns[col]} onChange={ev => {
              this.props.observable.trigger({key: "columns/" + col, value: ev.target.checked});
              this.setState({columns: Object.assign(this.state.columns, {[col]: ev.target.checked})});
            }} />
          {col}
        </label>
      ))}
      {this.renderDropdown("Log Filter", ["debug", "info", "warn", "err", "trace", "none"], lv => (
        <label>
          <input type="radio" checked={this.state.logLevel === lv} onChange={_ => {
              this.props.observable.trigger({key: "logLevel", value: lv});
              this.setState({logLevel: lv});
            }} />
          {lv}
        </label>
      ))}
      {this.renderDropdown("Set Log Level", ["debug", "info", "warn", "err", "trace", "button"], lv => {
        if (lv === "button") {
          return (<button
              style={{padding: "5px", marginLeft: "10px"}}
              onClick={_ => IrisLib.setLogLevel(this.state.setLogLevel)}>
                SET
            </button>)
        }
        else {
          return (<label>
              <input type="radio" checked={this.state.setLogLevel === lv} onChange={_ => {
                  this.setState({setLogLevel: lv});
                }} />
              {lv}
            </label>)
        }
      })}
    </div>
  );
  }
}

export default class Log {
  constructor() {
    this.view = View;
    this.name = "LOG";
    this.observable = IrisLib.createObservable();
    this.titleBar = <TitleForm observable={this.observable} />;
    this.layout = {
      x: 0, y: 0,
      w: 8, h: 6,
      minW: 6, maxW: 20,
      minH: 2, maxH: 20
    }
  }
}

File: dotnet/src/Frontend/fable/Frontend/Elmish/Log.fs
module Iris.Web.Log

open Fable.Import
open Fable.Helpers.React
open Fable.Helpers.React.Props
open Fable.Core
open Fable.Core.JsInterop
open Elmish.React
open Iris.Core
open Iris.Web.Core
open Helpers
open State
open Types

// fixed-data-table styles
importSideEffects("fixed-data-table/dist/fixed-data-table.css")

let Table: React.ComponentClass<obj> = importMember "fixed-data-table"
let Column: React.ComponentClass<obj> = importMember "fixed-data-table"
let Cell: React.ComponentClass<obj> = importMember "fixed-data-table"

module Columns =
  let [<Literal>] LogLevel = "LogLevel"
  let [<Literal>] Time = "Time"
  let [<Literal>] Tag = "Tag"
  let [<Literal>] Tier = "Tier"
  let [<Literal>] Message = "Message"
  let sortable = [LogLevel; Time; Tag; Tier]

let getWidth (col: string) =
  match col with
  | Columns.LogLevel -> 70
  | Columns.Time -> 80
  | Columns.Tag -> 75
  | Columns.Tier -> 75
  | Columns.Message -> 400
  | col -> failwithf "Unrecognized log column: %s" col

[<Emit("new Date($0).toLocaleTimeString()")>]
let printTime (t: uint32): string = jsNative

let readLog (col: string) (log: LogEvent) =
  match col with
  | Columns.LogLevel -> str (string log.LogLevel)
  | Columns.Time -> str (printTime log.Time)
  | Columns.Tag -> str log.Tag
  | Columns.Tier -> str (string log.Tier)
  | Columns.Message -> str log.Message
  | col -> failwithf "Unrecognized log column: %s" col

let onSorted dispatch (cfg: LogConfig) col _ =
  let direction =
    match cfg.sorting with
    | Some s when s.column = col ->
      s.direction.Reverse
    | _ -> Direction.Ascending
  { cfg with sorting = Some { column = col; direction = direction } }
  |> UpdateLogConfig |> dispatch

let sortableCell dispatch (cfg: LogConfig) col =
    let suffix =
      match cfg.sorting with
      | Some s when s.column = col ->
        match s.direction with
        | Ascending -> " ↓"
        | Descending -> " ↑"
      | _ -> ""
    from Cell
      %["style" => %["cursor" => "pointer"]
        "onClick" => onSorted dispatch cfg col]
      [str (col + " " + suffix)]

let makeSortableColumn dispatch cfg col =
  from Column
      %["width" => getWidth col
        "header" => sortableCell dispatch cfg col
        "cell" => fun (props: obj) ->
          from Cell props [readLog col cfg.viewLogs.[!!props?rowIndex]]
       ] []

let body dispatch model =
  let cfg = model.logConfig
  from Table
    %["rowsCount" => cfg.viewLogs.Length
      "rowHeight" => 30
      "headerHeight" => 30
      "width" => 700
      "height" => 600]
    [ yield! Columns.sortable |> Seq.choose (fun col ->
        match Map.tryFind col cfg.columns with
        | Some true -> makeSortableColumn dispatch cfg col |> Some
        | Some false | None -> None)
      yield from Column
        %["width" => getWidth Columns.Message
          "header" => from Cell %[] [str Columns.Message]
          "cell" => fun (props: obj) ->
            // %["style" => %["whiteSpace"=>"nowrap"]]
            from Cell props [readLog Columns.Message cfg.viewLogs.[!!props?rowIndex]]
         ] []
    ]

let dropdown title values generator =
  div [Class "iris-dropdown"] [
    div [Class "iris-dropdown-button"] [str title]
    div [Class "iris-dropdown-content"]
      (values |> List.map (fun x -> div [Key x] [generator x]))
  ]

let titleBar dispatch (model: Model) =
    let filter = defaultArg model.logConfig.filter ""
    div [] [
      input [
        Type "text"
        Placeholder "Filter by regex..."
        DefaultValue !^filter
        OnChange (fun ev ->
          let filter =
            match !!ev.target?value with
            | null | "" -> None
            | filter -> Some filter
          { model.logConfig with filter = filter } |> UpdateLogConfig |> dispatch)
      ]
      dropdown "Columns" Columns.sortable (fun col ->
        label [] [
          input [
            Type "checkbox"
            Checked model.logConfig.columns.[col]
            OnChange (fun ev ->
              let checked': bool = !!ev.target?``checked``
              let columns = Map.add col checked' model.logConfig.columns
              { model.logConfig with columns = columns } |> UpdateLogConfig |> dispatch)
          ]
          str col
        ]
      )
      dropdown "Log Filter" ["debug"; "info"; "warn"; "err"; "trace"; "none"] (fun strLv ->
        let lv =
          match strLv with
          | "none" -> None
          | lv -> Some(Iris.Core.LogLevel.Parse(lv))
        label [] [
          input [
            Type "radio"
            Checked (model.logConfig.logLevel = lv)
            OnChange (fun ev ->
              { model.logConfig with logLevel = lv }
              |> UpdateLogConfig |> dispatch)
          ]
          str strLv
        ]
      )
      dropdown "Set Log Level" ["debug"; "info"; "warn"; "err"; "trace"; "button"] (fun strLv ->
        if strLv = "button" then
          button [
            OnClick(fun _ ->
              model.logConfig.setLogLevel
              |> SetLogLevel
              |> ClientContext.Singleton.Post)
          ] [str "SET"]
        else
          let lv = Iris.Core.LogLevel.Parse(strLv)
          label [] [
            input [
              Type "radio"
              Checked (model.logConfig.setLogLevel = lv)
              OnChange (fun ev ->
                { model.logConfig with setLogLevel = lv }
                |> UpdateLogConfig |> dispatch)
            ]
            str strLv
          ])
    ]

let view id name dispatch model =
  widget id name (Some titleBar) body dispatch model

let createLogWidget(id: System.Guid) =
  { new IWidget with
    member __.Id = id
    member __.Name = "LOG"
    member __.InitialLayout =
      { i = id; ``static`` = false
        x = 0; y = 0
        w = 8; h = 6
        minW = 6; maxW = 20
        minH = 2; maxH = 20 }
    member this.Render(id, dispatch, model) =
      div [Key (string id)] [
        lazyViewWith
          (fun m1 m2 ->
              equalsRef m1.logs m2.logs
                  && equalsRef m1.logConfig m2.logConfig)
          (view id this.Name dispatch)
          model
      ]
  }
